tendermint/src/lite/verifier.rs

```diff
@@ -34,8 +32,64 @@ where
     Ok(())
 }
 
+// TODO: documentation!
+pub fn verify<S, V, L>(
+    sh1: &S,
+    h1_next_vals: &V,
+    sh2: &S,
+    h2_vals: &V,
+    trust_level: L,
+) -> Result<(), Error>
+where
+    S: SignedHeader,
+    V: ValidatorSet,
+    L: TrustThreshold,
+{
+    let h1 = sh1.header();
+    let h2 = sh2.header();
+    let commit = sh2.commit();
+    if h2.height() == h1.height().increment() {
+        if h2.validators_hash() != h1_next_vals.hash() {
+            return Err(Error::InvalidNextValidatorSet);
+        }
+    } else {
+        // ensure that +1/3 of last trusted validators signed correctly
+        if let Err(e) = verify_commit_trusting(h1_next_vals, commit, trust_level) {
+            return Err(e);
+        }
+    }
+
+    verify_header_and_commit(h2, sh2.commit(), h2_vals)
+}
+
+fn check_support<S, H, V, T, L>(
+    h1: &S,
+    h1_next_vals: &V,
+    h2: &S,
+    trust_threshold: L,
+    trusting_period: Duration,
+    now: SystemTime,
+) -> Result<(), Error>
+where
+    S: SignedHeader,
+    H: Header,
+    V: ValidatorSet,
+    L: TrustThreshold,
+{
+    if let Err(err) = expired(h1.header(), trusting_period, now) {
+        return Err(err);
+    }
+
+    if h2.header().height() == h1.header().height().increment()
+        && h2.header().validators_hash() != h1_next_vals.hash()
+    {
+        return Err(Error::InvalidNextValidatorSet);
+    }
+    verify_commit_trusting(h1_next_vals, h2.commit(), trust_threshold)
+}
+
```

Comment:
Looks good. Just a bit more cleanup maybe. It seems that `verify()` and `check_support()` have some dup code. Here is the summary of checks we do for a signed header:

1. is header properly formatted and signed -> this is `verify_header_and_commit()`. In order to check this we verify/ call:
   -  hashes: `validate_vals_and_commit()`
   -  more than 2/3rds have signed: `verify_commit_full()`
2. should the header be trusted -> this is `check_support()` and the checks are:
   - is the header expired?
   - should the validator set that signed the header be trusted?:  
       - for sequential we just check h1 next validator set is same as  h2 validator set. 
       - for skipping we check > 1/3rd have signed -> this is `verify_commit_trusting()`

`verify()` seems to put all these checks together so I think it should just call:
- `verify_header_and_commit()`
- `check_support()`

However, in the spec we do the two checks at different times, mainly because of the nature of bisection that comes back to already verified (as in 1. above) headers. In the spec:
- `verify()` performs 1.,  same as `verify_header_and_commit()` here
- `checkSupport()` performs 2, same as `check_support()` here

Some suggestions, maybe we could have some comments in the beginning of the file that describe this call flow. And maybe order the functions so the code is easier to read from top to bottom.

Reply:
Thanks @ancazamfir! Very useful feedback! 

Yeah, I was thinking of deleting the current existing `verify` function (which lends the naming and logic from the current go implementation) and follow the naming in the spec. I just need to update the tests to call `verify_header_and_commit()` and `check_support()` instead of the current `verify`. Then current `verify` can be deleted and name of `verify_header_and_commit` can be changed to `verify` which then should take a `SignedHeader` instead of (now) a separate `Header` and an associated `Commit`.  Passing in a SignedHeader instead makes the relation of the params clearer, too.

I spend quite some time on trying to get rid of the dependencies from the tendermint crate: Time, Id, Hash. Time was simply replaced by SystemTime. But with Id (which basically needs to implement the `Eq` trait) the mixture of pure generic functions on the one hand, and, traits with associated types (e.g. ValidatorId, ValidatorSet) and methods that return these, on the other hand (e.g. `fn validator(&self, val_id: Id) -> Option<Self::Validator>`) is a really interesting exercise but not an easy one (at least not for me). Almost felt like a very difficult puzzle / riddle :-D Will come back to this later. And do the changes you've mentioned now.

Reply:
Deleted `verify`, renamed `verify_header_and_commit` to now `verify`. Added a comment saying that you are expected to have called `verify` before calling calling `check_support`. More documentation coming later in the module documentation with an example (render using `cargo doc --open --no-deps`). Thanks again :-)

Reply:
Great. But won't `votingpower_in` check the signatures again redundantly? We can always address this later, but should track it.

Reply:
Yes, indeed. I've opened #96 to track feedback and Todos!